core/src/test/java/org/apache/iceberg/TestUpdatePartitionSpec.java

```diff
@@ -477,12 +477,12 @@ public void testAddRedundantTimePartition() {
   }
 
   @Test
-  public void testAddDeletedField() {
-    AssertHelpers.assertThrows("Should fail adding a duplicate field",
-        IllegalArgumentException.class, "Cannot add duplicate partition field",
-        () -> new BaseUpdatePartitionSpec(formatVersion, PARTITIONED)
-            .removeField("shard")
-            .addField(bucket("id", 16))); // duplicates shard
+  public void testNoEffectAddDeletedSameField() {
+    PartitionSpec updated = new BaseUpdatePartitionSpec(formatVersion, PARTITIONED)
+        .removeField("shard")
+        .addField(bucket("id", 16))
```

Comment:
This isn't correct because "shard" is a custom name. The result of separating this into two updates (remove and then add) would be a different partition spec, with the bucket field named `id_bucket_16`.

In this case, I think you just need to detect the rename and rewrite the remove/add to rename.

Reply:
it's a good point, updated